firedrake/dmplex.pyx

```diff
@@ -1083,3 +1056,816 @@ def plex_renumbering(PETSc.DM plex, np.ndarray[PetscInt, ndim=1, mode="c"] reord
+        return -1
+    elif x.remote_rank > y.remote_rank:
+        return 1
+
+    if x.global_u < y.global_u:
+        return -1
+    elif x.global_u > y.global_u:
+        return 1
+
+    if x.global_v < y.global_v:
+        return -1
+    elif x.global_v > y.global_v:
+        return 1
+
+    return 0
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline void get_communication_lists(
+    PETSc.DM plex, PETSc.Section vertex_numbering,
+    np.ndarray[np.int32_t, ndim=1, mode="c"] cell_ranks,
+    # Output parameters:
+    np.int32_t *nranks, np.int32_t **ranks, np.int32_t **offsets,
+    np.int32_t **facets, np.int32_t **facet2index):
+
+    """Creates communication lists for shared facet information exchange.
+
+    :arg plex: The DMPlex object encapsulating the mesh topology
+    :arg vertex_numbering: Section describing the universal vertex numbering
+    :arg cell_ranks: MPI rank of the owner of each (visible) non-owned cell,
+                     or -1 for (locally) owned cell.
+
+    :arg nranks: Number of neighbouring MPI nodes (return value)
+    :arg ranks: MPI ranks of neigbours (return value)
+    :arg offsets: Offset for each neighbour in the data buffer (return value)
+    :arg facets: Array of local plex facet numbers of shared facets
+                 (return value)
+    :arg facet2index: Maps local facet numbers to indices in the communication
+                      buffer, inverse of 'facets' (return value)
+    """
+    cdef:
+        int comm_size = MPI.comm.size
+        PetscInt cStart, cEnd
+        PetscInt nfacets, fStart, fEnd, f
+        PetscInt i, k, support_size
+        PetscInt *support = NULL
+        PetscInt local_count, remote
+        PetscInt v[2]
+        np.int32_t *facet_ranks = NULL
+        np.int32_t *nfacets_per_rank = NULL
+
+        CommFacet *cfacets = NULL
+
+    cStart, cEnd = plex.getHeightStratum(0)
+    fStart, fEnd = plex.getHeightStratum(1)
+    nfacets = fEnd - fStart
+
+    CHKERR(PetscMalloc1(nfacets, &facet_ranks))
+    memset(facet_ranks, -1, nfacets * sizeof(np.int32_t))
+
+    # Determines which facets are shared, and which MPI process
+    # they are shared with.
+    for f in range(fStart, fEnd):
+        CHKERR(DMPlexGetSupportSize(plex.dm, f, &support_size))
+        CHKERR(DMPlexGetSupport(plex.dm, f, &support))
+
+        local_count = 0
+        remote = -1
+        for i in range(support_size):
+            if cell_ranks[support[i] - cStart] >= 0:
+                remote = cell_ranks[support[i] - cStart]
+            else:
+                local_count += 1
+
+        if local_count == 1:
+            facet_ranks[f - fStart] = remote
+
+    # Counts how many facets are shared with each MPI node
+    CHKERR(PetscMalloc1(comm_size, &nfacets_per_rank))
+    memset(nfacets_per_rank, 0, comm_size * sizeof(np.int32_t))
+
+    for i in range(nfacets):
+        if facet_ranks[i] != -1:
+            nfacets_per_rank[facet_ranks[i]] += 1
+
```

Comment:
Probably not relevant for process numbers we're currently (and in the immediate future) likely to be running on.